src/NHibernate/Collection/Generic/PersistentGenericBag.cs

```diff
@@ -162,6 +167,26 @@ public bool IsReadOnly
 			get { return false; }
 		}
 
+		public Expression Expression
+		{
+			get { return InnerQueryable.Expression; }
+		}
+
+		public System.Type ElementType
+		{
+			get { return InnerQueryable.ElementType; }
+		}
+
+		public IQueryProvider Provider
+		{
+			get { return InnerQueryable.Provider; }
+		}
+
+		private IQueryable<T> InnerQueryable
+		{
+			get { return queryable ?? (queryable = new NhQueryable<T>(Session, this)); }
```

Comment:
`Lazy<>` is not really required. What's required is a proper implementation of `IQueryable<>`, and it turns out that `NhQueryable<>` is doing almost everything what we need.